extension/src/service/klarna-service.js

```diff
@@ -0,0 +1,86 @@
+const {
+  ADYEN_PERCENTAGE_MINOR_UNIT,
+  DEFAULT_PAYMENT_LANGUAGE,
+  KLARNA_DEFAULT_LINE_ITEM_NAME
+} = require('../config/klarna-constants')
+
+function createLineItems (payment, cart) {
+  const lineItems = []
+  const locales = _getLocales(cart, payment)
+
+  cart.lineItems.forEach((item) => {
+    lineItems.push(_createAdyenLineItemFromLineItem(item, locales))
+  })
+
+  cart.customLineItems.forEach((item) => {
+    lineItems.push(_createAdyenLineItemFromCustomLineItem(item, locales))
+  })
+
+  lineItems.push(_createShippingInfoAdyenLineItem(cart.shippingInfo))
+
+  return lineItems
+}
+
+function _getLocales (cart, payment) {
+  const locales = []
+  let paymentLanguage = payment.custom && payment.custom.fields['languageCode']
+  if (!paymentLanguage)
+    paymentLanguage = cart.locale
+  if (paymentLanguage)
+    locales.push(paymentLanguage)
+  if (!paymentLanguage || paymentLanguage !== DEFAULT_PAYMENT_LANGUAGE)
+    locales.push(DEFAULT_PAYMENT_LANGUAGE)
+
+  return locales
+}
+
+function _createAdyenLineItemFromLineItem (ctpLineItem, locales) {
+  return {
+    id: ctpLineItem.variant.sku,
+    quantity: ctpLineItem.quantity,
+    description: _localizeOrFallback(ctpLineItem.name, locales, KLARNA_DEFAULT_LINE_ITEM_NAME),
+    amountIncludingTax: ctpLineItem.price.value.centAmount,
+    taxPercentage: ctpLineItem.taxRate.amount * ADYEN_PERCENTAGE_MINOR_UNIT
```

Comment:
taxRate is an optional field, do we need to do a check here to avoid NPEs?

Reply:
Hm good point. I had a look in the docs and it says this:
```
Will be set automatically in the Platform TaxMode once the shipping address is set is set. For the External tax mode the tax rate has to be set explicitly with the ExternalTaxRateDraft.
```
So the question is does it make sense to check it here? When you make a payment, should you already have a cart with the shipping address already? Otherwise I would not make sense, right? Theoretically there could be NPE here, but practically I don't see a value for check. Maybe @butenkor can clarify here.

Reply:
If it's not possible you don't need to check it if it's possible it will block the process and the error will not be stored in interface interaction. Also how you will be sure that a cart has all the information that you need?

Reply:
You would need to try it out but according to docs, you can have a cart without shipping address but with `totalPrice` set (and no tax info on cart and line item level). Checking for NPE might be a good idea here.

Reply:
https://github.com/commercetools/commercetools-adyen-integration/pull/294/commits/e1fb9e8fec635a465b3b350798818a0c3279bee8